subsys/random/rand32_entropy_device.c

```diff
@@ -9,26 +9,18 @@
 #include <drivers/entropy.h>
 #include <string.h>
 
-static struct device *entropy_driver;
-
 #if defined(CONFIG_ENTROPY_DEVICE_RANDOM_GENERATOR)
 u32_t sys_rand32_get(void)
 {
-	struct device *dev = entropy_driver;
+	struct device *dev;
 	u32_t random_num;
 	int ret;
 
-	if (unlikely(!dev)) {
-		/* Only one entropy device exists, so this is safe even
-		 * if the whole operation isn't atomic.
-		 */
-		dev = device_get_binding(DT_CHOSEN_ZEPHYR_ENTROPY_LABEL);
-		__ASSERT((dev != NULL),
-			"Device driver for %s (DT_CHOSEN_ZEPHYR_ENTROPY_LABEL) not found. "
-			"Check your build configuration!",
-			DT_CHOSEN_ZEPHYR_ENTROPY_LABEL);
-		entropy_driver = dev;
-	}
+	dev = device_get_binding(DT_CHOSEN_ZEPHYR_ENTROPY_LABEL);
```

Comment:
hmm this isn't great for performance. device_get_binding() is unfortunately O(n) where n is the number of devices, and I think it does a string compare for each one. It's slow.

we could keep this as a global and put it in `z_libc_partition`...but this isn't ideal from a freedom-from-interference perspective as there's no way currently in Zephyr for different logical applications to get separate memory for library globals.

I suggest leaving `entropy_driver` caching in place if CONFIG_USERSPACE is disabled until we have some better solutions here, as this change will negatively affect non user mode

Reply:
Agree about performance, that was one of the reason I was reluctant about it.
I need to check with @jukkar what is the use case that he needs it in userspace, but I suspect it is related with moving network stack to userspace, maybe some high-level protocol requires it.
The other solution is make syscalls for these syscalls and then remove syscalls from entropy APIs (that I believe is not necessary).

Reply:
> but I suspect it is related with moving network stack to userspace, maybe some high-level protocol requires it.

Correct, I need this for example in IPv6 when network stack is running in user mode.

> The other solution is make syscalls for these syscalls and then remove syscalls from entropy APIs (that I believe is not necessary).

Would we need to grant the access if we would make the `sys_rand32_get()` a syscall, isn't that be a simpler for the user of this API too?

Reply:
yes. Actually looking other implementations like xoroshiro, that I initially though we wouldn't need access to a device we still have problems. All of the rely on globals to save  context.

I really think syscalls here will be much clear than have these globals into partitions that must be explicitly add to user threads memory domains + this api to grant permission to the used device.